### PR TITLE
feat(affiliates): 1776 add v2 dev mining based on tvl

### PR DESCRIPTION
**Pull request by daywiss**
Signed-off-by: David Adams <david@umaproject.org>

**Motivation**
https://app.clubhouse.io/uma-project/story/1776/update-the-dev-mining-script-to-use-tvl-instead-of-tvm-in-the-calculation

**Summary**

Adds tvl based rewarding as "v2" of devmining script. Keeps in place the original dev mining calculation but adds a second version based on collaterals only.  Much of the code is the same between the original version but some modifications.

Once merged the default unless specified otherwise will be to run v2. This should not be merged until after aug 2nd, since this is not needed until after.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [x]  Ran end-to-end test, running the code as in production
- [x]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

**Issue(s)**
https://app.clubhouse.io/uma-project/story/1776/update-the-dev-mining-script-to-use-tvl-instead-of-tvm-in-the-calculation

### CODE REVIEW

File: packages/affiliates/apps/DevMiningRewards.js [hunk 1]
```diff
@@ -101,6 +105,94 @@ const App = (env) => async (params) => {
   };
 };
 
+const v2 = async (env, params, DevMining) => {
```
Comment (daywiss) [inline packages/affiliates/apps/DevMiningRewards.js:108]: this is in the dev mining application. There are 2 blocks of code that configure things for v1 (tvm based dev mining) and now v2 (tvl based dev mining). This new block is almost a copy and paste of the v1 block, except it removes any mention of synthetic tokens.

File: packages/affiliates/libs/affiliates/devmining_v2.js [hunk 2]
```diff
@@ -0,0 +1,250 @@
+const highland = require("highland");
```
Comment (daywiss) [inline packages/affiliates/libs/affiliates/devmining_v2.js:1]: this is the devmining v2 business logic. Its almost completely identical to v1, but removes any calculations based on synthetic tokens, and just converts collateral to USD

File: packages/affiliates/libs/affiliates/devmining_v2.js [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+          // if EMP was expired this block, dont include it in further calculations.
+          if (isExpired) {
+            return result;
+          }
+          const [, closestCollateralPrice] = collateralTokenPrices[empIndex].closest(timestamp);
+          const totalCollateral = Object.values(collateral).reduce((result, value) => {
+            return result.add(toBN(value));
+          }, toBN("0"));
+          const value = calculateValue(totalCollateral, closestCollateralPrice, collateralTokenDecimals[empIndex]);
```
Comment (daywiss) [inline packages/affiliates/libs/affiliates/devmining_v2.js:133]: the bulk of the logic was just changed here, compared to v1. Calculate value is now just collateral value, and synthetics minted are ignored. Everything else for the most part is the same

File: packages/affiliates/libs/contracts.js [hunk 4]
```diff
@@ -129,6 +131,22 @@ function Emp({ abi = getAbi("ExpiringMultiParty"), web3 } = {}) {
...
+// amount: amount of token in wei
+// price: price as a float
+// decimals: decimal value for the token
+// returns value in wei 18 decimals
+function calculateValue(amount, price, decimals) {
+  price = toBN(toWei(price.toString()));
+  amount = toBN(amount);
+  return fromWei(amount.mul(price), decimals);
+}
+
```
Comment (daywiss) [inline packages/affiliates/libs/contracts.js:149]: a couple utility functions added. web3s `fromWei` is really wierd and does not work the same way as ethers, it requires you pass in a string like `wei` or `gwei` which is really useless, so i made a utility

File: packages/affiliates/scripts/basic-contract.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
+require("dotenv").config();
```
Comment (daywiss) [inline packages/affiliates/scripts/basic-contract.js:1]: can probably ignore this, or could be deleted, but could not run the app using `getWeb3`. would get `UnhandledPromiseRejectionWarning: Error: Invalid JSON RPC response: "" ` which went away once i initialized my own web3.

File: packages/affiliates/test/mocha/devmining-v2.js [hunk 6]
```diff
@@ -0,0 +1,321 @@
+const { DevMining } = require("../../libs/affiliates");
```
Comment (daywiss) [inline packages/affiliates/test/mocha/devmining-v2.js:1]: this copies the tests from v1, changes some values, but does a good job of testing scenarios

**Review by mrice32 — APPROVED** [review]
LGTM! This isn't super dry, but I'm not sure it's worth doing the work of trying to make it dry since this is not something we expect to be iterating on a bunch.

**Review by chrismaree — APPROVED** [review]

**Comment by daywiss** [comment]
> LGTM! This isn't super dry, but I'm not sure it's worth doing the work of trying to make it dry since this is not something we expect to be iterating on a bunch.

Thanks, yea i just didnt see the value in spending the extra time to try and reuse code, though you are right theres definatley a lot of duplicate code.

**Pull request merged**